### PR TITLE
add Series.str.remove(pre|suf)fix

### LINKED ISSUE #36944: ENH: add string method remove prefix and suffix, python 3.9
Opening this issue for discussion:

Since python 3.9 is out, and we have the new [string methods](https://www.python.org/dev/peps/pep-0616/) `removeprefix` and `removesuffix`, it would be nice to add them to the pandas string methods as well
```python
In [2]: import pandas as pd

In [3]: df = pd.DataFrame({'A': ['str_string1', 'str_string2', 'str_string3']})

In [4]: print(df)
             A
0  str_string1
1  str_string2
2  str_string3

In [5]: df['A'].str.removeprefix('str_')
Out[5]: 
0    string1
1    string2
2    string3
Name: A, dtype: object
```
An argument not to add this is that it's pretty easily to achieve with `str.split`:
```python
df['A'].str.split('_').str[-1]
```

### PR DESCRIPTION
**Pull request by janosh**
- [x] closes #36944
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Based on work in https://github.com/pandas-dev/pandas/pull/39226

### CODE REVIEW

File: pandas/tests/strings/test_strings.py
Comment (alimcmaster1) [inline pandas/tests/strings/test_strings.py:538]: Can we parameterize these tests cases using https://docs.pytest.org/en/6.2.x/parametrize.html

File: pandas/core/strings/object_array.py
Comment (alimcmaster1) [inline pandas/core/strings/object_array.py:417]: Can you add type hints for these newly added methods?

**Review by alimcmaster1 — COMMENTED** [review]
Looks like an excellent first PR. Could we also add benchmarks as suggested by @simonjayhawkins here: https://github.com/pandas-dev/pandas/pull/39226#issuecomment-836719770

Ref: https://pandas.pydata.org/docs/development/contributing_codebase.html#running-the-performance-test-suite

File: pandas/tests/strings/test_strings.py
Comment (janosh) [reply pandas/tests/strings/test_strings.py:538]: You mean like this?

```py
@pytest.mark.parametrize(
    "prefix,expected",
    [("x", ["xABCxx", "x BNSD", "LDFJH xx"]), ("xx ", ["xxABCxx", "BNSD", "LDFJH xx"])],
)
def test_removeprefix(any_string_dtype, prefix, expected):
    ser = Series(["xxABCxx", "xx BNSD", "LDFJH xx"], dtype=any_string_dtype)

    result = ser.str.removeprefix(prefix)
    ser_expected = Series(expected, dtype=any_string_dtype)
    tm.assert_series_equal(result, ser_expected)


@pytest.mark.parametrize(
    "suffix,expected",
    [("x", ["xxABCx", "xx BNSD", "LDFJH x"]), ("xx ", ["xxABCxx", "xx BNSD", "LDFJH"])],
)
def test_removesuffix(any_string_dtype, suffix, expected):
    ser = Series(["xxABCxx", "xx BNSD", "LDFJH xx"], dtype=any_string_dtype)

    result = ser.str.removesuffix(suffix)
    ser_expected = Series(expected, dtype=any_string_dtype)
    tm.assert_series_equal(result, ser_expected)
```

It's actually longer and I find it harder to read. But your call?

File: pandas/core/strings/object_array.py
Comment (janosh) [reply pandas/core/strings/object_array.py:417]: Just the inputs or return value as well?

File: pandas/tests/strings/test_strings.py
Comment (attack68) [reply pandas/tests/strings/test_strings.py:538]: Not quite sure why the strings need to be so complex. A minimal case is surely isomorphic..
```python

@pytest.mark.parametrize(
    "prefix, expected", [("a", ["b", " b c", "bc"]), ("ab", ["", "a b c", "bc"])],
)
def test_removeprefix(any_string_dtype, prefix, expected):
    ser = Series(["ab", "a b c", "bc], dtype=any_string_dtype)
    result = ser.str.removeprefix(prefix)
    ser_expected = Series(expected, dtype=any_string_dtype)
    tm.assert_series_equal(result, ser_expected)

@pytest.mark.parametrize(
    "suffix, expected", [("c", ["ab", "a b ", "b"]), ("bc", ["ab", "a b c", ""])],
)
def test_removesuffix(any_string_dtype, prefix, expected):
    ser = Series(["ab", "a b c", "bc"], dtype=any_string_dtype)
    result = ser.str.removesuffix(suffix)
    ser_expected = Series(expected, dtype=any_string_dtype)
    tm.assert_series_equal(result, ser_expected)
```

**Comment by janosh** [comment]
I tried `asv continuous -f 1.1 -E virtualenv upstream/master HEAD -b strings` but got

> · Unknown commit upstream/master

Would be good to add a hint to [the docs](https://pandas.pydata.org/docs/development/contributing_codebase.html#running-the-performance-test-suite) to run `git fetch upstream master` in that case.

Also, the following hint in `asv_bench/asv.conf.json` appears to be wrong:

> // The Pythons you'd like to test against.  If not provided, defaults
    // to the current version of Python used to run `asv`.
    // "pythons": ["2.7", "3.4"],
    "pythons": ["3.8"],

I got an error

> · No executable found for python 3.8
· No environments selected

Changing to `"3.9"` and rerunning, I get this error

> ·· Failure creating environment for virtualenv-py3.9-Cython0.29.21-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-scipy-sqlalchemy-tables-xlrd-xlsxwriter-xlwt
Traceback (most recent call last):
  File "/Users/janosh/.venv/py39/bin/asv", line 8, in <module>
    sys.exit(main())
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/main.py", line 38, in main
    result = args.func(args)
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/commands/__init__.py", line 49, in run_from_args
    return cls.run_from_conf_args(conf, args)
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/commands/continuous.py", line 75, in run_from_conf_args
    return cls.run(
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/commands/continuous.py", line 114, in run
    result = Run.run(
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/commands/run.py", line 294, in run
    Setup.perform_setup(environments, parallel=parallel)
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/commands/setup.py", line 89, in perform_setup
    list(map(_create, environments))
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/commands/setup.py", line 21, in _create
    env.create()
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/environment.py", line 704, in create
    self._setup()
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/plugins/virtualenv.py", line 148, in _setup
    self._install_requirements()
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/plugins/virtualenv.py", line 172, in _install_requirements
    self._run_pip(args, timeout=self._install_timeout, env=env)
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/plugins/virtualenv.py", line 177, in _run_pip
    return self.run_executable('python', ['-mpip'] + list(args), **kwargs)
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/environment.py", line 949, in run_executable
    return util.check_output([exe] + args, **kwargs)
  File "/Users/janosh/.venv/py39/lib/python3.9/site-packages/asv/util.py", line 754, in check_output
    raise ProcessError(args, retcode, stdout, stderr)
asv.util.ProcessError: Command '/Users/janosh/Repos/pandas/asv_bench/env/94b401bb918963c7d4d914702f131a3f/bin/python -mpip install -v --upgrade numpy Cython==0.29.21 matplotlib sqlalchemy scipy numba numexpr pyarrow tables openpyxl xlsxwriter xlrd xlwt odfpy jinja2' returned non-zero exit status 1

Does the Python interpreter have to be in a clean virtualenv or something like that?

File: pandas/core/strings/object_array.py
Comment (janosh) [reply pandas/core/strings/object_array.py]: None of the other methods have return value annotations. What would you like there?

```py
from typing import Any

def _str_removesuffix(self, suffix: str) -> np.ndarray[Any, np.str_]:
```

Or something simpler but not strictly correct?

```py
from typing import Sequence

def _str_removesuffix(self, suffix: str) -> Sequence[str]:
```

File: pandas/tests/strings/test_strings.py
Comment (janosh) [reply pandas/tests/strings/test_strings.py]: > Not quite sure why the strings need to be so complex.

Don't know either. I just copied those strings from the other test cases.

Comment (alimcmaster1) [reply pandas/tests/strings/test_strings.py]: Lets simplify them then - to what @attack68 suggested above.

Comment (janosh) [reply pandas/tests/strings/test_strings.py]: Yes, already have. Just waiting for feedback on return value type hints and perf benchmark before pushing.

**Comment by janosh** [comment]
I think I need some help here fixing this error:

```py
+ pytest -m 'not slow and not network and not clipboard' pandas --junitxml=test-data.xml
ImportError while loading conftest '/pandas/pandas/conftest.py'.
pandas/__init__.py:46: in <module>
    from pandas.core.api import (
pandas/core/api.py:29: in <module>
    from pandas.core.arrays import Categorical
pandas/core/arrays/__init__.py:7: in <module>
    from pandas.core.arrays.categorical import Categorical
pandas/core/arrays/categorical.py:113: in <module>
    from pandas.core.strings.object_array import ObjectStringArrayMixin
pandas/core/strings/__init__.py:31: in <module>
    from pandas.core.strings.base import BaseStringArrayMethods
pandas/core/strings/base.py:11: in <module>
    from pandas.core.series import Series
pandas/core/series.py:91: in <module>
    from pandas.core import (
pandas/core/generic.py:111: in <module>
    from pandas.core import (
pandas/core/indexing.py:57: in <module>
    from pandas.core.indexes.api import (
pandas/core/indexes/api.py:11: in <module>
    from pandas.core.indexes.base import (
pandas/core/indexes/base.py:128: in <module>
    from pandas.core.arrays import (
E   ImportError: cannot import name 'Categorical' from partially initialized module 'pandas.core.arrays' (most likely due to a circular import) (/pandas/pandas/core/arrays/__init__.py)
```

File: pandas/core/strings/base.py
Comment (simonjayhawkins) [inline pandas/core/strings/base.py:11]: import of Series is added just for the type annotations and causing the circular import?

try
```
from typing import TYPE_CHECKING
...
if TYPE_CHECKING:
    from pandas import Series
```

**Review by jreback — CHANGES_REQUESTED** [review]
can you add to docs:
- doc/source/reference/series.rst
- doc/source/user_guide/text.rst

merge master and ping on green

File: doc/source/whatsnew/v1.4.0.rst
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.4.0.rst:107]:
```suggestion
- Add :meth:`Series.str.removeprefix` and :meth:`Series.str.removesuffix` introduced in Python 3.9 to remove pre-/suffixes from string-type :class:`Series` (:issue:`36944`)
```

**Comment by janosh** [comment]
@jreback 3 tests timed out after 60 min. Is that equivalent to 'green' or anything still to do?

File: doc/source/user_guide/text.rst
Comment (jreback) [inline doc/source/user_guide/text.rst:338]: can you add a versionadded 1.4. tag here

Comment (jreback) [reply doc/source/user_guide/text.rst:338]: i think you need this instead of the one on L349

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment. ping on greenish

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @janosh very nice!

and @erfannariman for the original!

**Pull request merged**
